### DISCUSSION TITLE
User Account Duplication Based on Login Status and Email

### DISCUSSION THREAD

**Discussion by tsvetann** [medusajs/medusa#4102] [Feature Requests]
Hello,

I am experiencing an issue related to user account management that I believe could improve the overall user experience and streamline the system.

### Description of the issue:
When a new user registers and then logs in with their email address and makes an order, a new user account is automatically created. However, when the same user makes an order without logging in and uses the same email in the order details, a separate account is created. This results in two separate user accounts for the same individual, based on the same email address.

### Steps to reproduce:
Register and log in with an email address and place an order. A new user account is created.
Without logging in, place an order and provide the same email address in the order details. A separate user account is created.
Expected behavior:
Ideally, the system should recognize that the email address used to place the order (whether logged in or not) is the same, and therefore the orders should be linked to the same user account. This would prevent the creation of duplicate accounts and provide a more unified and consistent user experience.

### Actual behavior:
The system creates two separate user accounts for the same email address, one when the user is logged in and places an order, and another when the user places an order without logging in but provides the same email address.

### Proposed solution:
I suggest a mechanism that consolidates user accounts based on their email addresses. When an order is placed, the system could check whether an account with the provided email already exists. If it does, the order should be linked to the existing account, regardless of the user's login status at the time of the order. If no such account exists, a new one can be created.

This approach would help to:

- Improve user experience by providing a consolidated view of all orders under the same account (currently not the case).

- Reduce confusion and potential user frustration caused by multiple accounts.

- Streamline the management and maintenance of user accounts from an administrative perspective.

Thank you for considering this issue. I look forward to hearing your thoughts.

**Comment by pevey** [comment]
I think there is a way in the admin to move the order from the non-logged in account to the actual user account.

I think this is set up this way by design and is necessary to allow for guest checkout.  But it is optional.  The easiest way to avoid this issue (and what I do) is require the user to log in or create account before checkout.

If you let an order be made on a user account just by email and without having to log in, that would be a security issue.

If you want to have "guest checkout," you have to have this possibility that someone who has an account will opt not to log in and will do "guest checkout."

If you want to get a bit fancy, you could also dynamically check for an existing account on your frontend when the "guest" user enters their email address.  If there is an existing account, you could prompt the user to login.  This is what the default Magento theme does.  Personally, I think it is bad design for privacy reasons.  It leaks whether a given email address shops at a given web site.

**Reply by tsvetann** [reply]
Thanks for your reply @pevey . Do you know where I can find this functionality to merge anon orders to an existing account in the admin? On another note if it's already possible to check if an account exists via the non logged in API then isn't that already a security concern? Everyone who knows Medusa can then freely check directly in my public API if an account exists? Thanks

**Reply by pevey** [reply]
In the admin, on the main order page, in the Customer section, click the ... in the top right.  There is an option to Transfer Ownership.  I've never used it, but I assumed it might be for this type of situation.

**Comment by srindom** [comment]
The idea of having registered and unregistered customer entries is, exactly as @pevey points out, to avoid leaking customer details. If we always merged customer accounts you could have a scenario where someone could attempt to hijack an email - e.g. create an account for abc@email.com. If abc@email.com has existing orders you would now be able to see the order history. Likewise, if unregistered purchases got merged with a registered account you could have orders in your order history that you haven't made. 

Making the split ensures that guest checkouts are kept separate. You can transfer ownership like @pevey mentions, and there are also storefront APIs to enable customers to request an ownership transfer of a previous order.
